torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
     alignas(4096) {{input_t}} dequantized_B_buf[buf_size];
     {%- endif %}
 
-    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
-
-    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
-        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
-        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
-        b_bf16.store(dst);
-    };
-
-    auto load_B_tile = [&]({{input2_t}}* B_ptr, int idx, int num_b_rows) {
-        {{input_t}}* base_addr = dequantized_B_buf + idx;
-        {{kernel.unroll_pragma(8)}}
-        for (int i = 0; i < num_b_rows; i++) {
-            load_B_row(
-                B_ptr + i * b_tile_ptr_stride,
-                base_addr + i * 32
-            );
-        }
-    };
-    auto load_dequantized_B = [&](int n) {
+    auto load_dequantized_B = [&](int base_idx) {
         // Load a tile of B & cache it in L1D.
-        {{kernel.unroll_pragma(4)}}
-        for (int k = 0; k < K; k += {{block_k}}) {
-            int num_b_rows = (k < last_k_offset) ? 16 : tail_k_size;
-            {{kernel.unroll_pragma(2)}}
-            for (int tile_col = 0; tile_col <= 1; tile_col++) {
-                load_B_tile(
-                    const_cast<{{input2_t}}*>(B) + n + k * ldb + tile_col * {{16 * vnni_size}},
-                    (k / {{block_k // 2}} + tile_col) * num_elements_per_b_tile,
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n.
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
+        for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {
+        {%- if block_n in [16, 32] %}
+            auto b_int8 = at::vec::Vectorized<int8_t>::loadu(
+                base_addr + idx_q,
+                static_cast<int64_t>({{block_n}})
+            );
+            auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+            b_bf16.store(dequantized_B_buf + idx_dq);
```

Comment:
Shouldn't be `b_bf16.store(dequantized_B_buf + idx_dq, {{block_n}});`? If `{{block_n}}` is less than full vec lane, why we store number of full vec lane elements into dst?

Reply:
Thanks for catching this!

The opaque `at::vec::Vectorized` API is being used, and not intrinsics directly, I had wrongly assumed it'd automatically use AVX2 if 16 elements are read with `loadu`, but [you're right](https://github.com/pytorch/pytorch/blob/8b55f649610eab61ad3504bd0685da97dd95809e/aten/src/ATen/cpu/vec/vec512/vec512_int.h#L763) - `at::vec::Vectorized`  will use a ZMM register if AVX512 is supported.

Reply:
Not required anymore